pkg/http-tunnel/server/registry.go

```diff
@@ -0,0 +1,245 @@
+// Copyright (C) 2017 Michał Matczuk
+// Use of this source code is governed by an AGPL-style
+// license that can be found in the LICENSE file.
+
+package server
+
+import (
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"net"
+	"sync"
+)
+
+// RegistryItem holds information about hosts and listeners associated with a
+// client.
+type RegistryItem struct {
+	Hosts         []*HostAuth
+	Listeners     []net.Listener
+	ListenerNames []string
+	ClientName    string
+	ClientID      string
+}
+
+// HostAuth holds host and authentication info.
+type HostAuth struct {
+	Host string
+	Auth *Auth
+}
+
+type hostInfo struct {
+	identifier string
+	auth       *Auth
+}
+
+type registry struct {
+	source map[string]*RegistryItem //Origin Address based on host:port
+	items  map[string]*RegistryItem //Client name
+	hosts  map[string]*hostInfo
+	mu     sync.RWMutex
+	logger *log.Entry
+}
+
+func newRegistry(logger *log.Entry) *registry {
+	if logger == nil {
+		logger = log.NewEntry(log.StandardLogger())
+	}
+
+	return &registry{
+		items:  make(map[string]*RegistryItem),
+		source: make(map[string]*RegistryItem),
+		hosts:  make(map[string]*hostInfo),
+		logger: logger,
+	}
+}
+
+func (r *registry) PreSubscribe(origaddr string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if _, ok := r.source[origaddr]; ok {
+		r.logger.Errorf("error on pre-subscribe to registry this entry already exist %s", origaddr)
+		return
+	}
+	r.source[origaddr] = &RegistryItem{ClientID: origaddr}
+}
+
+// Subscribe allows to connect client with a given identifier.
+func (r *registry) Subscribe(cname string, origaddr string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if _, ok := r.items[cname]; ok {
+		r.logger.Errorf("error on subscribe to registry this entry already exist %s", origaddr)
+		return
+	}
+	reg := r.source[origaddr]
+	reg.ClientName = cname
+	//fmt.Printf("SUBSCRIBE REGISTRY Client Name: [%s] Client ID: [%s] value : %+v \n", cname, origaddr, reg)
+	r.items[cname] = reg
+	r.logger.WithFields(log.Fields{
+		"client-name": cname,
+		"client-id":   origaddr,
+		"data":        reg,
+	}).Info("REGISTRY SUBSCRIBE")
+}
+
+// GetID returns the ID for this client
+func (r *registry) GetID(cname string) string {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	v, ok := r.items[cname]
+	if !ok {
+		return ""
+	}
+	return v.ClientID
+}
+
+// Subscriber returns client identifier assigned to given host.
+func (r *registry) Subscriber(hostPort string) (string, *Auth, bool) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	h, ok := r.hosts[trimPort(hostPort)]
+	if !ok {
+		return "", nil, false
+	}
+	fmt.Printf("SUBSCRIBER REGISTRY [%s] value : %+v \n", hostPort, h)
```

Comment:
All of these commented and uncommented prints in this file should be logs or removed